docs/source/developer/extension_migration_2_3.rst

```diff
@@ -0,0 +1,115 @@
+``setup.py`` and ``pyproject.toml``.
+
+The upgrade script also updates the dependencies in ``package.json`` to the ``^3.0.0`` packages. Here is an example diff:
+
+.. code:: diff
+
+   index 6f1562f..3fcdf37 100644
+   --- a/package.json
+   +++ b/package.json
+   @@ -29,9 +29,13 @@
+      "scripts": {
+   -    "build": "tsc",
+   -    "build:labextension": "npm run clean:labextension && mkdirp myextension/labextension && cd myextension/labextension && npm pack ../..",
+   -    "clean": "rimraf lib tsconfig.tsbuildinfo",
+   +    "build": "jlpm run build:lib && jlpm run build:labextension:dev",
+   +    "build:prod": "jlpm run build:lib && jlpm run build:labextension",
+   +    "build:lib": "tsc",
+   +    "build:labextension": "jupyter labextension build .",
+   +    "build:labextension:dev": "jupyter labextension build --development True .",
+   +    "clean": "rimraf lib tsconfig.tsbuildinfo myextension/labextension",
+   +    "clean:all": "jlpm run clean:lib && jlpm run clean:labextension",
+      "clean:labextension": "rimraf myextension/labextension",
+      "eslint": "eslint . --ext .ts,.tsx --fix",
+      "eslint:check": "eslint . --ext .ts,.tsx",
+   @@ -59,12 +63,12 @@
+      ]
+      },
+      "dependencies": {
+   -    "@jupyterlab/application": "^2.0.0",
+   -    "@jupyterlab/apputils": "^2.0.0",
+   -    "@jupyterlab/observables": "^3.0.0",
+   +    "@jupyterlab/builder": "^3.0.0",
+   +    "@jupyterlab/application": "^3.0.0",
+   +    "@jupyterlab/apputils": "^3.0.0",
+   +    "@jupyterlab/observables": "^3.0.0",
+      "@lumino/algorithm": "^1.2.3",
+      "@lumino/commands": "^1.10.1",
+      "@lumino/disposable": "^1.3.5",
+   @@ -99,6 +103,13 @@
+   -    "typescript": "~3.8.3"
+   +    "typescript": "~4.0.1"
+      },
+      "jupyterlab": {
+   -    "extension": "lib/plugin"
+   +    "extension": "lib/plugin",
+   +    "outputDir": "myextension/labextension/"
+      }
+   }
+
+
+On the diff above, we see that additional development scripts are also added, as they are used by the new extension system workflow.
+
+The diff also shows the new ``@jupyterlab/builder`` as a ``devDependency``.
+``@jupyterlab/builder`` is a package required to build the extension as a federated extension.
+It hides away internal dependencies such as ``webpack``, and produces the assets that can then be distributed as part of a Python package.
+
+Extension developers do not need to interact with ``@jupyterlab/builder`` directly, but instead can use the
+``jupyter labextension build`` command. This command is run automatically as part of the ``build`` script
+(``jlpm run build``).
+
+Publishing the extension to PyPI and conda-forge
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Starting from JupyterLab 3.0, extensions can be distributed as a Python package.
+
+The extension tutorial provides explanations to package the extension so it can be
+published on PyPI and conda forge: :ref:`extension_tutorial_publish`.
```

Comment:
Should we also add that still npm packages might be needed in case an extension provides extension points?

Reply:
> npm packages might be needed in case an extension provides extension points?

@goanpeca I am intrigued here. I may have missed something in the issues/prs of the federated extension story. Can you link or shed light why npm would still be needed. My interpretation so far is that any extension would be pushed to pypi. Or are you referring to the core npm packages?

Reply:
So let's say you create a new extension that provides a new editor for files of type `X`, `DataLayerMirror`, now someone else might want to extend your extension, because `DataLayerMirror`  provides modes and extension points. 

To be able to develop for this and create the actual packages, you need to be able to install your extension as an NPM package, so you can develop locally and have typescript work correctly.

Does this clarify things a bit?

Reply:
Clear. I have not seen so far extensions extending extensions, but that is something in my head since some time.

Prolly the doc to write should put a few sentences on that concept like you have just done. I don't think there anything written on this, nor any example in the extension-example repository.

In that case, pushing the packages to npm.js makes indeed sense and is needed.

Reply:
> Should we also add that still npm packages might be needed in case an extension provides extension points?

Good point, we probably want to encourage extension authors to continue publishing to npm.

There are a couple of third-party extensions out there that build on top of each other such as [jupyterlab-topbar](https://github.com/jtpio/jupyterlab-topbar) and [jupyterlab-conda](https://github.com/mamba-org/jupyter_conda).

> it would be super helpfull to provide a "standard" example of using a github action to publish to both PyPI and NPM 

Maybe this should be part of the cookiecutter as an opt-in (https://github.com/jupyterlab/extension-cookiecutter-ts/issues/101). I guess it would also be fine to have a separate section in the JupyterLab documentation to cover this (if so can be addressed in a separate PR)

Reply:
Oh ye absolutely. I can make a separate PR for that :-)